hypershift-operator/ignitionserver/main.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+// This is the simplest http server that enable us to satisfy
+// 1 - 1 relation between clusters and ign endpoints.
+// At the moment this acts as purely a proxy towards MCS that are created on demand by NodePools,
+// this might / might not change in the future.
+// TODO (alberto): Support only https.
+// TODO (alberto): Support only token authenticated requests.
+// TODO (alberto): Parameterize listening port and URL to proxy.
+// TODO (alberto): Metrics.
+
+type ignHandler struct{}
+
+var _ http.Handler = &ignHandler{}
+
+func (h *ignHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// We only match /ign/$NodePoolName
+	ignPathPattern, err := regexp.Compile("^/ign/[^/ ]*$")
+	if err != nil {
+		panic(err)
+	}
+
+	log.Printf("User Agent %s", r.Header.Get("User-Agent"))
+	log.Printf("Requested %s", r.URL.Path)
+
+	if ignPathPattern.MatchString(r.URL.Path) {
+		urlPathSplit := strings.SplitAfter(r.URL.Path, "/")
+		// This can only happen if our regExp is not working as expected.
+		if len(urlPathSplit) > 3 {
+			panic(fmt.Sprintf("Unexpected URL path: %s", r.URL.Path))
+		}
+
+		targetUrl := fmt.Sprintf("http://machine-config-server-%s/config/master", urlPathSplit[2])
+
+		// Build proxy request.
+		proxyReq, err := http.NewRequest("GET", targetUrl, nil)
+		if err != nil {
+			panic(err)
+		}
+
+		// Copy original Headers into proxy request.
+		for k, vv := range r.Header {
+			for _, v := range vv {
+				proxyReq.Header.Add(k, v)
+			}
+		}
+
+		// Send proxy request.
+		client := &http.Client{}
+		resp, err := client.Do(proxyReq)
+		if err != nil {
+			// Send 500 response.
+			log.Printf("Server Error: %v", err)
+			http.Error(w, "Server Error", http.StatusInternalServerError)
+			return
+		}
+		defer resp.Body.Close()
+
+		// Copy headers into the responseWriter.
+		for k, vv := range resp.Header {
+			for _, v := range vv {
+				w.Header().Add(k, v)
+			}
+		}
+		w.WriteHeader(resp.StatusCode)
+
+		io.Copy(w, resp.Body)
+		return
+	}
+
+	// No pattern matched; send 404 response.
+	http.NotFound(w, r)
+}
+
+func main() {
+	r := &ignHandler{}
+	if err := http.ListenAndServe(":9090", r); err != nil {
```

Comment:
You can do graceful shutdown handling with a pattern like:

```go
ctx, cancel := context.WithCancel(context.Background())
sigs := make(chan os.Signal, 1)
signal.Notify(sigs, syscall.SIGINT)
go func() {
  <-sigs
  cancel()
}()

mux := http.NewServeMux()
mux.HandleFunc("/ignition", handler)

server := http.Server{Addr: addr, Handler: mux}

go func() {
  <-ctx.Done()
  if err := server.Shutdown(ctx); err != nil {
    log.Printf("error shutting down server: %s", err)
  }
}()

if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
  return err
}

```

Reply:
Thanks for sharing this! Changed to use this approach.